File: server/lib/jobs.ts
/**
 * Background jobs for long-running Supervisor calls.
 * Browser polls GET /api/job/:id until status is done|error.
 */

export type JobStatus = 'running' | 'done' | 'error';

export interface JobRecord {
  status: JobStatus;
  result: unknown;
  error: string;
  started: number;
  message?: string;
}

const jobs = new Map<string, JobRecord>();
const JOB_TTL_MS = 2 * 3600 * 1000;

function cleanupOldJobs(): void {
  const cutoff = Date.now() - JOB_TTL_MS;
  for (const [id, job] of jobs) {
    if (job.started < cutoff) jobs.delete(id);
  }
}

export function newJobId(): string {
  cleanupOldJobs();
  return crypto.randomUUID();
}

export function createJob(id: string, message?: string): void {
  jobs.set(id, { status: 'running', result: null, error: '', started: Date.now(), message });
}

export function updateJobMessage(id: string, message: string): void {
  const job = jobs.get(id);
  if (job?.status === 'running') {
    jobs.set(id, { ...job, message });
  }
}

export function finishJob(id: string, result: unknown): void {
  const job = jobs.get(id);
  if (!job) return;
  jobs.set(id, { ...job, status: 'done', result, error: '' });
}

export function failJob(id: string, error: string): void {
  const job = jobs.get(id);
  if (!job) return;
  jobs.set(id, { ...job, status: 'error', result: null, error });
}

export function getJob(id: string): JobRecord | undefined {
  return jobs.get(id);
}

export function runningJobCount(): number {
  return [...jobs.values()].filter(j => j.status === 'running').length;
}

export function loadStatus(): { state: string; step: number; elapsed: number } {
  const running = [...jobs.values()].filter(j => j.status === 'running');
  if (!running.length) {
    const done = [...jobs.values()].filter(j => j.status === 'done');
    if (done.length) return { state: 'done', step: 3, elapsed: 0 };
    return { state: 'idle', step: 0, elapsed: 0 };
  }
  const oldest = running.reduce((a, b) => (a.started < b.started ? a : b));
  const elapsed = Math.floor((Date.now() - oldest.started) / 1000);
  let step = 1;
  if (elapsed >= 5) step = 2;
  if (elapsed >= 15) step = 3;
  return { state: 'running', step, elapsed };
}

File: py_server/lib/jobs.py
"""Background jobs for long-running Supervisor calls — mirrors server/lib/jobs.ts."""
from __future__ import annotations

import time
import uuid
from typing import Any, Literal, TypedDict

JobStatus = Literal["running", "done", "error"]


class JobRecord(TypedDict, total=False):
    status: JobStatus
    result: Any
    error: str
    started: float
    message: str


_jobs: dict[str, JobRecord] = {}
JOB_TTL_MS = 2 * 3600 * 1000


def _cleanup_old_jobs() -> None:
    cutoff = (time.time() * 1000) - JOB_TTL_MS
    for job_id, job in list(_jobs.items()):
        if job["started"] < cutoff:
            del _jobs[job_id]


def new_job_id() -> str:
    _cleanup_old_jobs()
    return str(uuid.uuid4())


def create_job(job_id: str, message: str | None = None) -> None:
    record: JobRecord = {
        "status": "running",
        "result": None,
        "error": "",
        "started": time.time() * 1000,
    }
    if message is not None:
        record["message"] = message
    _jobs[job_id] = record


def update_job_message(job_id: str, message: str) -> None:
    job = _jobs.get(job_id)
    if job and job.get("status") == "running":
        _jobs[job_id] = {**job, "message": message}


def finish_job(job_id: str, result: Any) -> None:
    job = _jobs.get(job_id)
    if not job:
        return
    _jobs[job_id] = {**job, "status": "done", "result": result, "error": ""}


def fail_job(job_id: str, error: str) -> None:
    job = _jobs.get(job_id)
    if not job:
        return
    _jobs[job_id] = {**job, "status": "error", "result": None, "error": error}


def get_job(job_id: str) -> JobRecord | None:
    return _jobs.get(job_id)


def running_job_count() -> int:
    return sum(1 for job in _jobs.values() if job.get("status") == "running")


def load_status() -> dict[str, int | str]:
    running = [job for job in _jobs.values() if job.get("status") == "running"]
    if not running:
        done = [job for job in _jobs.values() if job.get("status") == "done"]
        if done:
            return {"state": "done", "step": 3, "elapsed": 0}
        return {"state": "idle", "step": 0, "elapsed": 0}
    oldest = min(running, key=lambda j: j["started"])
    elapsed = int((time.time() * 1000 - oldest["started"]) // 1000)
    step = 1
    if elapsed >= 5:
        step = 2
    if elapsed >= 15:
        step = 3
    return {"state": "running", "step": step, "elapsed": elapsed}
